File: app/src/main/java/com/cronoslot/CameraActivity.java
package com.cronoslot;

import android.app.*;
import android.content.*;
import android.graphics.*;
import android.media.*;
import android.os.*;
import android.view.*;
import android.widget.*;
import androidx.activity.ComponentActivity;
import androidx.camera.core.*;
import androidx.camera.lifecycle.ProcessCameraProvider;
import androidx.camera.view.PreviewView;
import androidx.core.content.ContextCompat;
import java.util.*;
import java.util.concurrent.*;


public class CameraActivity extends ComponentActivity {
 PreviewView preview; LineOverlay overlay; TextView info,lapsView; Db db; ExecutorService exec=Executors.newSingleThreadExecutor();
 byte[] previous; boolean armed=true,calibrationOnly=false; long lastDetection=0,lastLapAt=0,raceStart=0; List<Double> lapTimes=new ArrayList<>();
 long pilotId,carId,trackId; String remote=""; ToneGenerator tone; android.content.SharedPreferences prefs;

 @Override public void onCreate(Bundle b){super.onCreate(b);db=new Db(this);pilotId=getIntent().getLongExtra("pilotId",0);carId=getIntent().getLongExtra("carId",0);trackId=getIntent().getLongExtra("trackId",0);remote=getIntent().getStringExtra("remote");calibrationOnly=getIntent().getBooleanExtra("calibrationOnly",false);prefs=getSharedPreferences("calibration",0);build();startCamera();}
 void build(){FrameLayout root=new FrameLayout(this);preview=new PreviewView(this);root.addView(preview,new FrameLayout.LayoutParams(-1,-1));overlay=new LineOverlay(this);root.addView(overlay,new FrameLayout.LayoutParams(-1,-1));overlay.lineX=prefs.getFloat("lineX_"+trackId,.5f);
  LinearLayout panel=new LinearLayout(this);panel.setOrientation(LinearLayout.VERTICAL);panel.setPadding(14,14,14,14);panel.setBackgroundColor(0xBB000000);info=new TextView(this);info.setTextColor(Color.WHITE);info.setTextSize(17);lapsView=new TextView(this);lapsView.setTextColor(Color.WHITE);lapsView.setTextSize(23);lapsView.setText("Vueltas: 0");panel.addView(info);panel.addView(lapsView);
  if(calibrationOnly){info.setText("Mueve la línea roja hasta el punto de paso.");panel.addView(btn("GUARDAR CALIBRACIÓN",v->{prefs.edit().putFloat("lineX_"+trackId,overlay.lineX).apply();Toast.makeText(this,"Calibración guardada.",Toast.LENGTH_SHORT).show();finish();}));}
  else{info.setText("Pasa el coche por la línea roja.");panel.addView(btn("ARMAR DETECCIÓN",v->{armed=true;info.setText("DETECCIÓN ACTIVA");}));panel.addView(btn("🔴 TERMINAR CARRERA",v->confirmFinish()));}
  FrameLayout.LayoutParams pp=new FrameLayout.LayoutParams(-1,-2);pp.gravity=Gravity.TOP;root.addView(panel,pp);setContentView(root);tone=new ToneGenerator(AudioManager.STREAM_NOTIFICATION,95);
 }
 Button btn(String t,View.OnClickListener l){Button b=new Button(this);b.setText(t);b.setOnClickListener(l);return b;}
 void startCamera(){ProcessCameraProvider.getInstance(this).addListener(()->{try{ProcessCameraProvider p=ProcessCameraProvider.getInstance(this).get();Preview pr=new Preview.Builder().build();pr.setSurfaceProvider(preview.getSurfaceProvider());ImageAnalysis an=new ImageAnalysis.Builder().setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST).setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_YUV_420_888).build();an.setAnalyzer(exec,this::analyze);p.unbindAll();p.bindToLifecycle(this,CameraSelector.DEFAULT_BACK_CAMERA,pr,an);}catch(Exception e){runOnUiThread(()->info.setText("Error cámara: "+e.getMessage()));}},ContextCompat.getMainExecutor(this));}
 void analyze(ImageProxy img){try{ByteBufferView b=new ByteBufferView(img.planes[0].getBuffer());byte[] data=b.data();int w=img.getWidth(),h=img.getHeight();int x=(int)(overlay.lineX*w);double score=0;int n=0;if(previous!=null&&previous.length==data.length){for(int y=(int)(h*.2);y<(int)(h*.8);y+=4)for(int xx=Math.max(0,x-14);xx<Math.min(w,x+15);xx+=3){int i=y*w+xx;if(i>=0&&i<data.length){score+=Math.abs((data[i]&255)-(previous[i]&255));n++;}}}previous=data;double s=n==0?0:score/n;long now=SystemClock.elapsedRealtime();if(!calibrationOnly&&armed&&s>22&&now-lastDetection>250){lastDetection=now;armed=false;runOnUiThread(()->{if(lastLapAt==0){raceStart=now;info.setText("CRONÓMETRO INICIADO");}else{double sec=(now-lastLapAt)/1000.0;lapTimes.add(sec);lapsView.setText("Vuelta "+lapTimes.size()+" · "+String.format(Locale.getDefault(),"%.3f s",sec));recordTone(sec);}lastLapAt=now;});}else if(!armed&&s<9)armed=true;}catch(Exception ignored){}finally{img.close();}}
 void recordTone(double sec){Double bt=db.bestTrack(trackId),bc=db.bestCombo(trackId,pilotId,carId);boolean tr=bt==null||sec<bt,co=bc==null||sec<bc;if(co){tone.startTone(ToneGenerator.TONE_PROP_BEEP,140);if(tr){new Handler(Looper.getMainLooper()).postDelayed(()->tone.startTone(ToneGenerator.TONE_PROP_BEEP,140),190);info.setText("🏆 DOBLE RÉCORD");}else info.setText("🏎️ RÉCORD DE COCHE");}else if(tr){tone.startTone(ToneGenerator.TONE_PROP_BEEP,140);info.setText("🏆 RÉCORD DE PISTA");}}
 void confirmFinish(){new AlertDialog.Builder(this).setTitle("¿Realmente quieres terminar la sesión?").setMessage("Se guardarán todos los tiempos registrados.").setNegativeButton("NO, CONTINUAR",null).setPositiveButton("SÍ, TERMINAR",(d,w)->askNotes()).show();}
 void askNotes(){EditText e=new EditText(this);e.setHint("Notas de la sesión (opcional)");new AlertDialog.Builder(this).setTitle("Notas de la sesión").setMessage("Puedes anotar neumáticos u otras condiciones de esta prueba.").setView(e).setNeutralButton("OMITIR",(d,w)->save("")) .setPositiveButton("GUARDAR SESIÓN",(d,w)->save(e.getText().toString())).show();}
 void save(String notes){if(lapTimes.isEmpty()){Toast.makeText(this,"No hay vueltas registradas.",Toast.LENGTH_SHORT).show();finish();return;}db.addSession(pilotId,carId,trackId,remote,raceStart,notes,lapTimes);Toast.makeText(this,"Sesión guardada.",Toast.LENGTH_LONG).show();finish();}
 @Override protected void onDestroy(){if(exec!=null)exec.shutdown();if(tone!=null)tone.release();super.onDestroy();}
 static class ByteBufferView{java.nio.ByteBuffer buf;ByteBufferView(java.nio.ByteBuffer b){buf=b;}byte[] data(){byte[] d=new byte[buf.remaining()];buf.get(d);return d;}}
 static class LineOverlay extends View{float lineX=.5f;Paint p=new Paint(1);Paint h=new Paint(1);LineOverlay(Context c){super(c);p.setColor(Color.RED);p.setStrokeWidth(7);h.setColor(Color.WHITE);}protected void onDraw(Canvas c){super.onDraw(c);float x=getWidth()*lineX;c.drawLine(x,0,x,getHeight(),p);c.drawCircle(x,getHeight()/2f,18,h);}public boolean onTouchEvent(MotionEvent e){if(e.getAction()==MotionEvent.ACTION_DOWN||e.getAction()==MotionEvent.ACTION_MOVE){lineX=Math.max(.03f,Math.min(.97f,e.getX()/getWidth()));invalidate();return true;}return true;}}
}

File: app/src/main/java/com/cronoslot/CameraActivity.kt
package com.cronoslot
import android.content.Context
import android.graphics.*
import android.media.*
import android.os.*
import android.view.*
import android.widget.*
import androidx.activity.ComponentActivity
import androidx.appcompat.app.AlertDialog
import androidx.camera.core.*
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import java.util.concurrent.Executors
import kotlin.math.abs

class CameraActivity:ComponentActivity(){
 private lateinit var pv:PreviewView;private lateinit var line:LineView;private lateinit var info:TextView;private lateinit var list:TextView
 private val ex=Executors.newSingleThreadExecutor();private var prev:ByteArray?=null;private var armed=true;private var last=0L;private var lapAt=0L;private val laps=mutableListOf<Double>();private val tone=ToneGenerator(AudioManager.STREAM_NOTIFICATION,90)
 private val cooldown=300L
 override fun onCreate(b:Bundle?){super.onCreate(b);ui();camera()}
 private fun ui(){val root=FrameLayout(this);pv=PreviewView(this);root.addView(pv,FrameLayout.LayoutParams(-1,-1));line=LineView(this);root.addView(line,FrameLayout.LayoutParams(-1,-1));val p=LinearLayout(this).apply{orientation=LinearLayout.VERTICAL;setPadding(12,12,12,12);setBackgroundColor(0xB0000000.toInt())};info=TextView(this).apply{text="Mueve la línea roja hasta la meta";setTextColor(Color.WHITE);textSize=16f};list=TextView(this).apply{text="Vueltas: 0";setTextColor(Color.WHITE);textSize=22f};p.addView(info);p.addView(list);p.addView(Button(this).apply{text="ARMAR DETECCIÓN";setOnClickListener{armed=true;info.text="DETECCIÓN ACTIVA"}});p.addView(Button(this).apply{text="🔴 TERMINAR CARRERA";setOnClickListener{finishRace()}});val pp=FrameLayout.LayoutParams(-1,-2);pp.gravity=Gravity.TOP;root.addView(p,pp);setContentView(root)}
 private fun camera(){val f=ProcessCameraProvider.getInstance(this);f.addListener({val c=f.get();val pr=Preview.Builder().build().also{it.surfaceProvider=pv.surfaceProvider};val an=ImageAnalysis.Builder().setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST).build();an.setAnalyzer(ex){a(it)};c.unbindAll();c.bindToLifecycle(this,CameraSelector.DEFAULT_BACK_CAMERA,pr,an)},ContextCompat.getMainExecutor(this))}
 private fun a(img:ImageProxy){try{val b=img.planes[0].buffer;val d=ByteArray(b.remaining());b.get(d);val w=img.width;val h=img.height;val x=(line.xpos*w).toInt().coerceIn(12,w-13);var score=0.0;var n=0;prev?.let{o->if(o.size==d.size){for(y in (h*.25).toInt() until (h*.75).toInt() step 5)for(xx in x-10..x+10 step 3){val i=y*w+xx;if(i<d.size){score+=abs((d[i].toInt()and 255)-(o[i].toInt()and 255));n++}}}};prev=d;val s=if(n==0)0.0 else score/n;val now=SystemClock.elapsedRealtime();if(armed&&s>22&&now-last>cooldown){last=now;armed=false;runOnUiThread{if(lapAt>0){val sec=(now-lapAt)/1000.0;laps.add(sec);tone.startTone(ToneGenerator.TONE_PROP_BEEP,120);list.text="Vuelta ${laps.size}: %.3f s".format(sec)}else info.text="Primera detección — cronómetro iniciado";lapAt=now}}else if(!armed&&s<9)armed=true}catch(_:Exception){}finally{img.close()}}
 private fun finishRace(){AlertDialog.Builder(this).setTitle("¿Realmente quieres terminar?").setMessage("Se guardarán los tiempos registrados.").setNegativeButton("CANCELAR",null).setPositiveButton("TERMINAR"){_,_->notes()}.show()}
 private fun notes(){val e=EditText(this);e.hint="Notas de la sesión (opcional)";AlertDialog.Builder(this).setTitle("Notas de la sesión").setView(e).setNegativeButton("OMITIR"){_,_->finish()}.setPositiveButton("GUARDAR SESIÓN"){_,_->Toast.makeText(this,"Sesión preparada con ${laps.size} vueltas.",Toast.LENGTH_LONG).show();finish()}.show()}
 override fun onDestroy(){ex.shutdown();tone.release();super.onDestroy()}
 class LineView(c:Context):View(c){var xpos=.5f;private val p=Paint(1).apply{color=Color.RED;strokeWidth=7f};override fun onDraw(c:Canvas){val x=width*xpos;c.drawLine(x,0f,x,height.toFloat(),p)};override fun onTouchEvent(e:MotionEvent):Boolean{if(e.action==0||e.action==2){xpos=(e.x/width).coerceIn(.03f,.97f);invalidate();return true};return true}}
}
